cdc_iterator.go

```diff
@@ -0,0 +1,293 @@
+	"github.com/go-mysql-org/go-mysql/canal"
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/schema"
+	"gopkg.in/tomb.v2"
+)
+
+type cdcIterator struct {
+	canal *canal.Canal
+	acks  *csync.WaitGroup
+	data  chan *canal.RowsEvent
+
+	// canalTomb is a tomb that simplifies handling canal run errors
+	canalTomb *tomb.Tomb
+
+	config cdcIteratorConfig
+}
+
+type cdcIteratorConfig struct {
+	common.SourceConfig
+	position  sdk.Position
+	TableKeys common.TableKeys
+}
+
+func newCdcIterator(ctx context.Context, config cdcIteratorConfig) (common.Iterator, error) {
+	c, err := common.NewCanal(config.SourceConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create canal: %w", err)
+	}
+
+	sdk.Logger(ctx).Info().Msg("created canal")
+
+	iterator := &cdcIterator{
+		canal:     c,
+		acks:      &csync.WaitGroup{},
+		data:      make(chan *canal.RowsEvent),
+		canalTomb: &tomb.Tomb{},
+		config:    config,
+	}
+
+	startPosition, err := iterator.getStartPosition(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get start position: %w", err)
+	}
+
+	iterator.canalTomb.Go(func() error {
+		ctx := iterator.canalTomb.Context(ctx)
+		return iterator.runCanal(ctx, startPosition)
+	})
+
+	// close the data channel when all tomb goroutines are done, which will happen only when
+	// the iterator teardown method is called
+	go func() {
+		<-iterator.canalTomb.Dead()
+		close(iterator.data)
+	}()
+
+	return iterator, nil
+}
+
+func (c *cdcIterator) getStartPosition(config cdcIteratorConfig) (mysql.Position, error) {
+	if config.position != nil {
+		pos, err := common.ParseSDKPosition(config.position)
+		if err != nil {
+			return mysql.Position{}, fmt.Errorf("failed to parse position: %w", err)
+		}
+		if pos.Kind == common.PositionTypeSnapshot {
+			return mysql.Position{}, fmt.Errorf("invalid position type: %s", pos.Kind)
+		}
+
+		return pos.CdcPosition.Position, nil
+	}
+
+	masterPos, err := c.canal.GetMasterPos()
+	if err != nil {
+		return mysql.Position{}, fmt.Errorf("failed to get master position: %w", err)
+	}
+
+	return masterPos, nil
+}
+
+func (c *cdcIterator) runCanal(ctx context.Context, startPos mysql.Position) error {
+	// buffered to allow the goroutine to exit when the iterator is torn down
+	errChan := make(chan error, 1)
+	go func() {
+		handler := &cdcEventHandler{
+			ctx:  ctx,
+			data: c.data,
+		}
+		c.canal.SetEventHandler(handler)
+		errChan <- c.canal.RunFrom(startPos)
+	}()
+
+	select {
+	case <-ctx.Done():
+		c.canal.Close()
+		return fmt.Errorf("context cancelled from runCanal: %w", ctx.Err())
+	case err := <-errChan:
+		return fmt.Errorf("failed to run canal: %w", err)
+	}
+}
+
+func (c *cdcIterator) Ack(context.Context, sdk.Position) error {
+	c.acks.Done()
+	return nil
+}
+
+func (c *cdcIterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, fmt.Errorf("context cancelled: %w", ctx.Err())
```

Comment:
Nitpick: wrapping a context error with `"context canceled"` doesn't give use additional info. In fact, the context error contains exactly the same message if it gets canceled.